toolchain/check/cpp/type_mapping.cpp

```diff
@@ -164,6 +184,25 @@ static auto TryMapClassType(Context& context, SemIR::ClassType class_type)
     case SemIR::RecognizedTypeInfo::CppVoidBase: {
       return ast_context.VoidTy;
     }
+    case SemIR::RecognizedTypeInfo::Optional: {
+      auto args = context.inst_blocks().GetOrEmpty(type_info.args_id);
+      if (args.size() == 1) {
+        auto arg_id = args[0];
+        if (auto facet = context.insts().TryGetAs<SemIR::FacetValue>(arg_id)) {
+          arg_id = facet->type_inst_id;
+        }
+        if (auto pointer_type =
+                context.insts().TryGetAs<SemIR::PointerType>(arg_id)) {
+          return WrappedType{
+              .inner_type_id = context.types().GetTypeIdForTypeInstId(
+                  pointer_type->pointee_id),
+              .wrap_fn = [](Context& context, clang::QualType inner_type) {
+                return context.ast_context().getPointerType(inner_type);
+              }};
```

Comment:
This is a `function_ref` to a lambda that outlives the lambda itself... In this case, maybe its ok because this is a stateless lambda? But I'm not sure?

If this is OK, will anything catch if this stops being OK? Could we rename this te `wrap_fn_ref` or something to make it less surprising when this doesn't work?

Or if this is not OK, maybe `std::function` and then `wrap_fn` seems good?